File: Clinica-Backend/src/main/java/com/pe/laboratorio/reports/audit/service/AuditService.java
package com.pe.laboratorio.reports.audit.service;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.stereotype.Service;

import com.pe.laboratorio.reports.audit.entity.AuditLog;
import com.pe.laboratorio.reports.audit.repository.AuditRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AuditService {

    private final AuditRepository auditRepository;

    public void logAction(String module, String action, String details,
            String username, Long userId, String ipAddress,
            String status, String userAgent, String url, String method) {

        AuditLog log = AuditLog.builder()
                .module(module)
                .action(action)
                .details(details)
                .username(username)
                .userId(userId)
                .ipAddress(ipAddress)
                .status(status)
                .userAgent(userAgent)
                .url(url)
                .method(method)
                .timestamp(LocalDateTime.now())
                .build();
        auditRepository.save(log);
    }

    public List<AuditLog> getAllLogs() {
        return auditRepository.findAllByOrderByTimestampDesc();
    }

    public List<AuditLog> getLogsByUser(String username) {
        return auditRepository.findByUsernameOrderByTimestampDesc(username);
    }

    /**
     * Búsqueda avanzada para la bitácora
     */
    public List<AuditLog> searchLogs(String username, String action,
            LocalDateTime startDate, LocalDateTime endDate) {

        // Manejo de strings vacíos como null para ignorar filtro
        String userFilter = (username != null && !username.trim().isEmpty()) ? username : null;
        String actionFilter = (action != null && !action.equals("Todas")) ? action : null;

        return auditRepository.searchLogs(userFilter, actionFilter, startDate, endDate);
    }
}

File: Clinica-frontend/src/app/features/reports/services/audit.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface AuditLog {
    id: number;
    userId: number;
    username: string;
    module: string;
    action: string;
    details: string;
    ipAddress: string;
    status: string;
    userAgent: string;
    url: string;
    method: string;
    timestamp: string;
}

@Injectable({
    providedIn: 'root'
})
export class AuditService {
    private apiUrl = 'http://localhost:8080/api/reports/audit';

    constructor(private http: HttpClient) { }

    getLogs(filters: any = {}): Observable<AuditLog[]> {
        let params = new HttpParams();

        if (filters.username) params = params.set('username', filters.username);
        if (filters.action && filters.action !== 'Todas') params = params.set('action', filters.action);
        if (filters.startDate) params = params.set('startDate', filters.startDate);
        if (filters.endDate) params = params.set('endDate', filters.endDate);

        return this.http.get<AuditLog[]>(this.apiUrl, { params });
    }
}
